ably/realtime_conn.go

```diff
@@ -84,6 +84,10 @@ func newConn(opts *ClientOptions, auth *Auth, callbacks connCallbacks) (*Connect
 }
 
 func (c *Connection) dial(proto string, u *url.URL) (proto.Conn, error) {
+	// (RTN23b)
+	query := u.Query()
+	query.Add("heartbeats", "true")
```

Comment:
RTN23b says that you may rely on the WebSocket's heartbeat mechanism, which we [already do by setting the connection's read deadline](https://github.com/ably/ably-go/blob/e362d92f93b0b04408c1b4299ae0ea34a0d64f93/ably/internal/ablyutil/websocket.go#L25). So it shouldn't be necessary to send the `heartbeat=true` param.

Reply:
I have checked our websocket library to see how it handles the Ping message.

https://github.com/golang/net/blob/ab34263943818b32f575efc978a3d24e80b04bd7/websocket/hybi.go#L291-L303

See there, this only receives ping and sends pong message without resetting the deadline.
Meaning the deadline which we only set once on every Receive call applies to all `Read` on conn and hence it will timeout even though we received a Ping from ably.

In short the deadline is never reset after receiving ping, so reads will eventually timeout based on the deadline set once on every Receive.

There is no api for custom handling of ping for this library so its impossible to rely on websocket pings

Reply:
The deadline is resetted when the connection receives _any_ data (which is the behavior we want), not at the `net/websocket` layer, but at the underlying `net` layer.

Reply:
> 
	// SetDeadline sets the read and write deadlines associated
	// with the connection. It is equivalent to calling both
	// SetReadDeadline and SetWriteDeadline.
	//
	// A deadline is an absolute time after which I/O operations
	// fail with a timeout (see type Error) instead of
	// blocking. The deadline applies to all future and pending
	// I/O, not just the immediately following call to Read or
	// Write. After a deadline has been exceeded, the connection
	// can be refreshed by setting a deadline in the future.
	
@tcard documentation for `net.Conn` says otherwise. All future calls will timeout when deadline is reached, you need to manually call `SetDeadline` again to reset it.

So, it isn't reset at all. Reads from `Ping` won't reset it.

Reply:
You're right, sorry for the confusion. I agree we should activate Ably heartbeats. We could go lower-level and handle incoming frames ourselves but I don't think it's worth the complexity.